dash/include/dash/halo/Halo.h

```diff
@@ -0,0 +1,1104 @@
+  typename PatternT::index_type {
+  return last - first;
+}
+
+template <typename ElementT, typename PatternT, dim_t NumDimensions>
+class Region {
+ private:
+  using GlobMemT =
+    GlobStaticMem<ElementT, dash::allocator::SymmetricAllocator<ElementT>>;
+  using HaloRegionSpecT = HaloRegionSpec<NumDimensions>;
+  using ViewSpecT       = typename PatternT::viewspec_type;
+  using region_index_t  = typename HaloRegionSpecT::region_index_t;
+  using size_type       = typename PatternT::size_type;
+  using BorderT         = std::array<bool, NumDimensions>;
+
+ public:
+  using iterator       = RegionIter<ElementT, PatternT>;
+  using const_iterator = const iterator;
+
+ public:
+  Region( const HaloRegionSpecT& region_spec, const ViewSpecT& region,
+          GlobMemT& globmem, const PatternT& pattern, const BorderT& border )
+  : _region_spec( region_spec ), _region( region ), _border( border ),
+    _border_region(
+      std::any_of( border.begin(), border.end(),
+                   []( bool border_dim ) { return border_dim == true; } ) ),
+    _beg( globmem, pattern, _region, 0, _region.size() ),
+    _end( globmem, pattern, _region, _region.size(), _region.size() ) {}
+
+  const region_index_t index() const { return _region_spec.index(); }
+
+  const HaloRegionSpecT& regionSpec() const { return _region_spec; }
+
+  const ViewSpecT& region() const { return _region; }
+
+  constexpr size_type size() const { return _region.size(); }
+
+  constexpr BorderT border() const { return _border; }
+
+  bool borderRegion() const { return _border_region; };
+
+  constexpr bool borderDim( dim_t dim ) const { return _border[dim]; }
+
+  iterator begin() const { return _beg; }
+
+  iterator end() const { return _end; }
+
+ private:
+  const HaloRegionSpecT _region_spec;
+  const ViewSpecT       _region;
+  BorderT               _border;
+  bool                  _border_region;
+  iterator              _beg;
+  iterator              _end;
+};  // Region
+
+template <typename ElementT, typename PatternT>
+class HaloBlock {
+ private:
+  static constexpr dim_t NumDimensions = PatternT::ndim();
+
+  using SelfT = HaloBlock<ElementT, PatternT>;
+  using GlobMemT =
+    GlobStaticMem<ElementT, dash::allocator::SymmetricAllocator<ElementT>>;
+  using CycleSpecT      = CycleSpec<NumDimensions>;
+  using index_type      = typename PatternT::index_type;
+  using ViewSpecT       = typename PatternT::viewspec_type;
+  using halo_reg_spec_t = HaloSpec<NumDimensions>;
+  using HaloRegionSpecT = HaloRegionSpec<NumDimensions>;
+  using region_t        = Region<ElementT, PatternT, NumDimensions>;
+  using regions_t       = std::vector<region_t>;
+  using halo_extent_t   = typename halo_reg_spec_t::extent_t;
+  using HaloExtsMaxT =
+    std::array<std::pair<halo_extent_t, halo_extent_t>, NumDimensions>;
+  using size_type     = typename PatternT::size_type;
+  using CoordsT       = std::array<index_type, NumDimensions>;
+  using RegionCoordsT = RegionCoords<NumDimensions>;
```

Comment:
Type naming is inconsistent: `SelfT` vs `size_type` vs `region_t`. I know we're sloppy throughout DASH with this but let's try to be consistent within one class at least  (helps readability) :)